[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/IsmailAkram/Assignment_4/blob/main/Assignment_4.ipynb)

In [1]:
# Imports and pip installations (if needed)
import pandas as pd
import numpy as np

Context reference for future reviews: https://hmcnich.gitbook.io/csc448-f21/assignment-4-judging-flowers

# Part 1: Load the dataset

In [2]:
# Load the dataset (load remotely, not locally)
from sklearn.datasets import load_iris
iris = load_iris()
iris_df = pd.DataFrame(iris['data'], columns = iris.feature_names) # transforming load_iris dictionary into a dataset
iris_df["species"] = iris.target # adding species column

In [3]:
# Output the first 15 rows of the data
iris_df.head(15)

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                 5.1               3.5                1.4               0.2   
1                 4.9               3.0                1.4               0.2   
2                 4.7               3.2                1.3               0.2   
3                 4.6               3.1                1.5               0.2   
4                 5.0               3.6                1.4               0.2   
5                 5.4               3.9                1.7               0.4   
6                 4.6               3.4                1.4               0.3   
7                 5.0               3.4                1.5               0.2   
8                 4.4               2.9                1.4               0.2   
9                 4.9               3.1                1.5               0.1   
10                5.4               3.7                1.5               0.2   
11                4.8               3.4                1.6               0.2   
12                4.8               3.0                1.4               0.1   
13                4.3               3.0                1.1               0.1   
14                5.8               4.0                1.2               0.2   

    species  
0         0  
1         0  
2         0  
3         0  
4         0  
5         0  
6         0  
7         0  
8         0  
9         0  
10        0  
11        0  
12        0  
13        0  
14        0

In [4]:
# Display a summary of the table information (number of datapoints, etc.)
iris_df.info() # for technical summary
iris_df.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
 4   species            150 non-null    int32  
dtypes: float64(4), int32(1)
memory usage: 5.4 KB


sepal length (cm)  sepal width (cm)  petal length (cm)  \
count         150.000000        150.000000         150.000000   
mean            5.843333          3.057333           3.758000   
std             0.828066          0.435866           1.765298   
min             4.300000          2.000000           1.000000   
25%             5.100000          2.800000           1.600000   
50%             5.800000          3.000000           4.350000   
75%             6.400000          3.300000           5.100000   
max             7.900000          4.400000           6.900000   

       petal width (cm)     species  
count        150.000000  150.000000  
mean           1.199333    1.000000  
std            0.762238    0.819232  
min            0.100000    0.000000  
25%            0.300000    0.000000  
50%            1.300000    1.000000  
75%            1.800000    2.000000  
max            2.500000    2.000000

In [5]:
iris_df.size # number of datapoints

750

## About the dataset
Explain what the data is in your own words. What are your features and labels? What is the mapping of your labels to the actual classes?

This is a multivariable data set exploring linear discrimmination analysis. It deals with sepal length, sepal width, petal length, petal width and 3 species of Iris. (Paraphrased the Wikipedia article).

Our features are the aforementioned "sepal length (cm)", "sepal width (cm)", "petal length (cm)", and "petal width (cm)"; and our label is "species".

Mapping: so we're looking at each feature and their values to determine what species they best match up with.

For example, in the first row we have: Sepal length: 5.1, Sepal width: 3.5, Petal length: 1.4, Petal width: 0.2 which determines that our species is "0" or "setosa".

# Part 2: Split the dataset into train and test

In [6]:
# Take the dataset and split it into our features (X) and label (y)
features = iris_df[["sepal length (cm)", "sepal width (cm)", "petal length (cm)", "petal width (cm)"]] # X (2D array)
label = iris_df["species"] # y

In [7]:
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(features, label, test_size=0.1, train_size=0.9) # splits into 4 arrays

In [8]:
X_train, X_test, y_train, y_test # for display purposes

(     sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
 53                 5.5               2.3                4.0               1.3
 149                5.9               3.0                5.1               1.8
 66                 5.6               3.0                4.5               1.5
 52                 6.9               3.1                4.9               1.5
 136                6.3               3.4                5.6               2.4
 ..                 ...               ...                ...               ...
 38                 4.4               3.0                1.3               0.2
 113                5.7               2.5                5.0               2.0
 68                 6.2               2.2                4.5               1.5
 132                6.4               2.8                5.6               2.2
 45                 4.8               3.0                1.4               0.3
 
 [135 rows x 4 columns],
      sepal length (cm)  

# Part 3: Logistic Regression

In [9]:
# i. Use sklearn to train a LogisticRegression model on the training set
from sklearn.linear_model import LogisticRegression

LogReg = LogisticRegression(max_iter = 500).fit(X_train, y_train) # for fitting; use train. *limit on max inters

In [10]:
# ii. For a sample datapoint, predict the probabilities for each possible class
pd.DataFrame(LogReg.predict_proba([[1, 2, 3, 4]]), columns = ["Setosa", "Versicolor", "Virginica"]) # for the entire X_test

# the probabilities for each class are the respective 3 columns.

Setosa  Versicolor  Virginica
0  0.315912    0.098862   0.585226

In [11]:
# But now I'll test sample values from Wikipedia's chart to see if they correlerate with the correct species.
LogReg.predict(np.array([[4.9, 3.1, 1.5, 0.1]])) # 0 meaning Setosa

array([0])

In [12]:
LogReg.predict(np.array([[4.9, 2.4, 3.3, 1.0]])) # 1 meaning Versicolor

array([1])

In [13]:
LogReg.predict(np.array([[6.4, 2.8, 5.6, 2.1]])) # 2 meaning Virginica

array([2])

In [14]:
# iii. Report on the score for Logistic regression model, what does the score measure?
LogReg.score(X_test, y_test) # for scoring; use test.

0.8666666666666667

The 0, 1, and 2 from the predictions correspond to Setosa, Versicolor, and Virginica respectively. 

The 0.8666666666666667 score means there's an 86% chance of this classification being correct.

In [15]:
# iv. Extract the coefficents and intercepts for the boundary line(s)
LogReg.coef_

array([[-0.52148695,  0.85762961, -2.43999629, -1.05024718],
       [ 0.38738272, -0.34925058, -0.23493114, -0.7556851 ],
       [ 0.13410423, -0.50837903,  2.67492743,  1.80593228]])

In [16]:
LogReg.intercept_

array([ 10.50894184,   3.05864558, -13.56758742])

# Part 4: Support Vector Machine

In [17]:
# i. Use sklearn to train a Support Vector Classifier on the training set
from sklearn import svm

# model = svm.SVC(kernel = "linear")
# model.fit(X_train, y_train)
# svm.SVC(LogReg, X_train, y_train)

svm_model = svm.SVC(probability = True).fit(features, label)

In [18]:
# ii. For a sample datapoint, predict the probabilities for each possible class
pd.DataFrame(svm_model.predict_proba([[1, 2, 3, 4]]), columns = ["Setosa", "Versicolor", "Virginica"])  # for the entire X_test

# the probabilities for each class are the respective 3 columns.

Setosa  Versicolor  Virginica
0  0.251848    0.237385   0.510768

In [19]:
# Like before, I'll test sample values from Wikipedia's chart to see if they correlerate with the correct species.
svm_model.predict(np.array([[4.9, 3.1, 1.5, 0.1]])) # 0 meaning Setosa

array([0])

In [20]:
svm_model.predict(np.array([[4.9, 2.4, 3.3, 1.0]])) # 1 meaning Versicolor

array([1])

In [21]:
svm_model.predict(np.array([[6.4, 2.8, 5.6, 2.1]])) # 2 meaning Virginica

array([2])

In [22]:
# iii. Report on the score for the SVM, what does the score measure?
svm_model.score(X_test, y_test)

0.9333333333333333

The 0, 1, and 2 from the predictions correspond to Setosa, Versicolor, and Virginica respectively. 

The 0.9333333333333333 score means there's an 93% chance of this classification being correct.

# Part 5: Neural Network

In [23]:
# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set
from sklearn.neural_network import MLPClassifier

mlp_model_lbfgs = MLPClassifier(solver='lbfgs', alpha=1e-5, hidden_layer_sizes=(15,), random_state=1).fit(features, label)

In [24]:
# ii. For a sample datapoint, predict the probabilities for each possible class
pd.DataFrame(mlp_model_lbfgs.predict_proba([[1, 2, 3, 4]]), columns = ["Setosa", "Versicolor", "Virginica"])

Setosa  Versicolor     Virginica
0     0.0         1.0  9.018592e-15

In [25]:
# Like before, I'll test sample values from Wikipedia's chart to see if they correlerate with the correct species.
mlp_model_lbfgs.predict(np.array([[4.9, 3.1, 1.5, 0.1]])) # 0 meaning Setosa

array([0])

In [26]:
mlp_model_lbfgs.predict(np.array([[4.9, 2.4, 3.3, 1.0]])) # 1 meaning Versicolor

array([1])

In [27]:
mlp_model_lbfgs.predict(np.array([[6.4, 2.8, 5.6, 2.1]])) # 2 meaning Virginica

array([2])

In [28]:
# iii. Report on the score for the Neural Network, what does the score measure?
mlp_model_lbfgs.score(X_test, y_test)

1.0

The 0, 1, and 2 from the predictions correspond to Setosa, Versicolor, and Virginica respectively.

The 1 score means there's an 100% chance of this classification being correct.

In [29]:
# iv: Experiment with different options for the neural network, report on your best configuration (the highest score I was able to achieve was 0.8666)
mlp_model_sgd = MLPClassifier(max_iter = 1500, solver='sgd', alpha=1e-5, hidden_layer_sizes=(15,), random_state=1).fit(features, label)

In [30]:
pd.DataFrame(mlp_model_sgd.predict_proba([[1, 2, 3, 4]]), columns = ["Setosa", "Versicolor", "Virginica"])

Setosa  Versicolor  Virginica
0  0.000004    0.000076   0.999921

In [38]:
mlp_model_sgd.score(X_test, y_test) # I got 1 in all three cases.

1.0

In [32]:
mlp_model_adam = MLPClassifier(max_iter = 1550, solver='adam', alpha=1e-5, hidden_layer_sizes=(15,), random_state=1).fit(features, label)

In [33]:
pd.DataFrame(mlp_model_adam.predict_proba([[1, 2, 3, 4]]), columns = ["Setosa", "Versicolor", "Virginica"])

Setosa  Versicolor  Virginica
0  0.000001    0.000001   0.999998

In [34]:
mlp_model_adam.score(X_test, y_test)

1.0

# Part 6: K-Nearest Neighbors

In [35]:
# i. Use sklearn to 'train' a k-Neighbors Classifier
# Note: KNN is a nonparametric model and technically doesn't require training
# fit will essentially load the data into the model see link below for more information
# https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier
from sklearn.neighbors import KNeighborsClassifier

KNN_model = KNeighborsClassifier(n_neighbors=3).fit(X_train, y_train) # for fitting; use train. *limit on max inters

In [36]:
# ii. For a sample datapoint, predict the probabilities for each possible class
pd.DataFrame(KNN_model.predict_proba([[1, 2, 3, 4]]), columns = ["Setosa", "Versicolor", "Virginica"])

Setosa  Versicolor  Virginica
0     0.0         1.0        0.0

In [37]:
# iii. Report on the score for kNN, what does the score measure?
KNN_model.score(X_test, y_test)

0.9333333333333333

The 0.9333333333333333 score means there's an 93% chance of this classification being correct.

# Part 7: Conclusions and takeaways

In your own words describe the results of the notebook. Which model(s) performed the best on the dataset? Why do you think that is? Did anything surprise you about the exercise?

Logistic: 
probabilties 0.315912	0.098862	0.585226
score: 0.8666666666666667

SVM: 
probabilities: 0.251848	0.237385	0.510768
score: 0.9333333333333333

Neural Network
lbfgs
probabilties: 0.0	1.0	9.018592e-15
score: 1

sgd
probabilties: 0.000004	0.000076	0.999921
score: 1

adam
probabilities: 0.000001	0.000001	0.999998
score: 1

KNN
probabilities: 0.0	1.0	0.0
score: 0.9333333333333333

It seems like Neural Networks performed the best on the dataset.

But I've also different scores upon each rerun from all models netting the same score (either all 1s, 86%, and 93% etc).

I assume this is due to reshuffling as I reran my code several times and used max_iter where appropriate, sometimes I got all 1s which is possibility but doesn't seem right? I would also get 86% and now I've got 93% I'll just leave this as is.

For Part 5 iv. you said that the highest score you got was 86%, I got that as well during one of the runs, but now I have 100%.

The main surprise is the aforementioned shuffling and my scores being sporadic.